cmd/ndm_daemonset/probe/eventhandler.go

```diff
@@ -110,3 +171,51 @@ func (pe *ProbeEvent) initOrErrorEvent() {
 		klog.Error(err)
 	}
 }
+
+// generateUUID creates a new UUID based on the algorithm proposed in
+// https://github.com/openebs/openebs/pull/2666
+func generateUUID(bd blockdevice.BlockDevice) (string, bool) {
+	var ok bool
+	var uuidField, uuid string
+
+	// select the field which is to be used for generating UUID
+	//
+	// Serial number is not used for UUID generation. This is because serial number is not
+	// unique in some cloud environments. For example, in GCP the serial number is
+	// configurable by the --device-name flag while attaching the disk.
+	// If this flag is not provided, GCP automatically assigns the serial number
+	// which is unique only to the node.
+	//
+	// PartitionTableUUID is not used for UUID generation in NDM. The only case where the disk has a PartitionTable
+	// and not partition is when, the user has manually created a partition table without writing any actual partitions.
+	// This means NDM will have to give its consumers the entire disk, i.e consumers will have access to the sectors
+	// where partition table is written. If consumers decide to reformat or erase the disk completely the partition
+	// table UUID is also lost, making NDM unable to identify the disk. Hence, even if a partition table is present
+	// NDM will rewrite it and create a new GPT table and a single partition. Thus consumers will have access only to
+	// the partition and the unique data will be stored in sectors where consumers do not have access.
+
+	switch {
+	case bd.DeviceAttributes.DeviceType == libudevwrapper.UDEV_PARTITION:
+		// The partition entry UUID is used when a partition (/dev/sda1) is processed. The partition UUID should be used
+		// if available, other than the partition table UUID, because multiple partitions can have the same partition table
+		// UUID, but each partition will have a different UUID.
+		klog.Infof("device(%s) is a partition, using partition UUID: %s", bd.DevPath, bd.PartitionInfo.PartitionEntryUUID)
+		uuidField = bd.PartitionInfo.PartitionEntryUUID
+		ok = true
+	case len(bd.DeviceAttributes.WWN) > 0:
+		klog.Infof("device(%s) has a WWN, using WWN: %s", bd.DevPath, bd.DeviceAttributes.WWN)
+		uuidField = bd.DeviceAttributes.WWN
+		ok = true
+	case len(bd.FSInfo.FileSystemUUID) > 0:
```

Comment:
what is the behavior on the Gke with local attached SSD? That will have file system, and later if we wipe that fs, will we deactivate that device?

Reply:
BlockDevices with filesystem should be given to the user only when the VolumeMode in BDC is `FileSystem`. 
Still if the user wipes the filesystem, on next restart we will mark that block device as inactive. 
The restart is required because NDM currently only listens for udev events, which are not generated when filesystem is wiped.